### ISSUE TITLE
zmath translation produces weirdly rotated translation matrix which causes a translation in the vertex shader to fail

### ISSUE THREAD

**Issue by KekOnTheWorld** [michal-z/zig-gamedev#268]
When directly passing the translation matrix generated by this code into a vertex shader

```zig
pub fn translation(x: f32, y: f32, z: f32) Mat {
    return .{
        f32x4(1.0, 0.0, 0.0, 0.0),
        f32x4(0.0, 1.0, 0.0, 0.0),
        f32x4(0.0, 0.0, 1.0, 0.0),
        f32x4(x, y, z, 1.0),
    };
}
```

and then multiplying it with the position unexpected behavior occurs.

By switching the translation matrix with (Taken from https://www.euclideanspace.com/maths/algebra/matrix/functions/inverse/index.htm)

```zig
const mat = Mat{
    f32x4(1.0, 0.0, 0.0, x),
    f32x4(0.0, 1.0, 0.0, y),
    f32x4(0.0, 0.0, 1.0, z),
    f32x4(0, 0, 0, 1.0),
};
```

the translation works as expected.

**Comment by michal-z** [comment]
This works as expected. `zmath` uses row-major matrices, row vectors (each row vector is stored in a SIMD register). You basically need to transpose the matrix before sending it to the shader. Please see our sample applications.

**Comment by KekOnTheWorld** [comment]
Okay, thanks for your comment

**Comment by KekOnTheWorld** [comment]
The translation matrix is now no longer an issue, I am having troubles with zm.orthographicRh tho. I used orthographicRh to generate a screen projection matrix for my 2d projection of lines,  I do not know if I am using this as intended.

```zig
return zm.transpose(zm.orthographicRh(camera.width, camera.height, 1, 10));
```

This statement alone causes my object to disappear from the screen space.

I found that setting the fourth value of the third row to zero solves this issue:

```zig
mat[2][3] = 0.0;
```

This value seams to be evaluated through the near and far plane:

```zig
const r = 1 / (near - far);
return .{
    f32x4(2 / w, 0.0, 0.0, 0.0),
    f32x4(0.0, 2 / h, 0.0, 0.0),
    f32x4(0.0, 0.0, r, 0.0),
    f32x4(0.0, 0.0, r * near, 1.0
};
 ```
 
Do you mind explaining what I am doing incorrectly? Should I even be using orthographicRh for 2D projections? I could just switch to zm.scaling to transform for screen width/height. Would this be a viable solution?
 
 Thanks in advance.

**Comment by hazeycode** [comment]
@KekOnTheWorld some thing that would be useful to note:
- What graphics API are you using?
- Do you have backface culling on?
- Are your vertices CW or CCW wound?
- Do you want Y-up or Y-down?
- Are you using pre or post multiplication in your shader?

**Comment by hazeycode** [comment]
@KekOnTheWorld I think I spotted the issue. You are setting the near to 1.0 here: `zm.orthographicRh(camera.width, camera.height, 1, 10)` did you mean 0.0? NOte that `mat[2][3] = 0.0;` effectively sets it to 0.0

**Comment by KekOnTheWorld** [comment]
> 

Ay, yeah setting the near plane to 0 fixes the issue. I am new to this. Thank you.

Will defo take a look at render doc tho.

**Comment by michal-z** [comment]
@KekOnTheWorld Can we close this issue?

**Comment by KekOnTheWorld** [comment]
Yep. This works now. Thanks for your help!

**Comment by presentfactory** [comment]
This should be re-opened I think as I just hit this problem as well. Zmath is supposed to be row-major and storing each row of a matrix linearly in memory, but if you look at the `zmath.translation` function it is clearly laying out the matrix column my column, making it column major:
```
return .{
    f32x4(1.0, 0.0, 0.0, 0.0),
    f32x4(0.0, 1.0, 0.0, 0.0),
    f32x4(0.0, 0.0, 1.0, 0.0),
    f32x4(x, y, z, 1.0),
};
```
A translation matrix takes the form:
```
1 0 0 X
0 1 0 Y
0 0 1 Z
0 0 0 1
```
So row by row it should be:
```
return .{
    f32x4(1.0, 0.0, 0.0, x),
    f32x4(0.0, 1.0, 0.0, y),
    f32x4(0.0, 0.0, 1.0, z),
    f32x4(0.0, 0.0, 0.0, 1.0),
};
```

I am unsure what zmath's actual convention here is. It provides both row and column major multiplication functions so it's hard to tell. Maybe it's row major in intent but with column major memory ordering? This should be clarified or fixed if it's a mistake.

**Comment by michal-z** [comment]
`zmath` uses the same convention as `DirectXMath`:

"DirectXMath uses row-major matrices, row vectors, and pre-multiplication. Handedness is determined by which function version is used (RH vs. LH), otherwise the function works with either left-handed or right-handed view coordinates."

Translation matrix in `zmath` is correct. Vectors are stored row-by-row (`row0`, `row1`, `row2`, `row3`) and translation coordinates are stored in `row3`. When you use pre-multiplication (`vector` * `matrix`) you will get correctly transformed *row* vector. You can also use post-multiplication (`matrix` * `vector`) and you will get correctly transformed *column* vector.

In other words, translation matrix looks exactly the same in raw memory in both cases because row-major and pre-multiplication gives you the same result as column-major and post-multiplication just transposed (result is a row vector in the first case and column vector in the second case).

**Comment by presentfactory** [comment]
Yeah but that's not what I am doing. I am saying if you simply call `zmath.translation` and do nothing with it other than upload it to OpenGL it is a column major matrix, because it is laid out column by column with respect to what a translation matrix is. The math may be doing it row major wise but the memory layout is column major and it's not mentioned anywhere.

**Comment by michal-z** [comment]
> Yeah but that's not what I am doing. I am saying if you simply call `zmath.translation` and do nothing with it other than upload it to OpenGL it is a column major matrix, because it is laid out column by column with respect to what a translation matrix is. The math may be doing it row major wise but the memory layout is column major and it's not mentioned anywhere.

No, memory layout is row major. Please re-read my answer.

**Comment by presentfactory** [comment]
Yeah I guess the usual mathematical matrix form is column major by default and that is just transposed for row major ones, so laying it out row major is the same as laying it out via column major for a column major matrix.

This stuff is always confusing I guess. The original response to this was wrong though as you do not need to transpose it when sending it to the shader since row and column major matrices will have the same memory layout.

**Comment by hazeycode** [comment]
> The original response to this issue was wrong though as you do not need to transpose it when sending it to the shader since row and column major matrices will have the same memory layout.

No, if your shader expects column-major matrices, as is the default for GLSL, then you should transpose before uploading.

Michal was pointing out that a row-major matrix can transform both row and column vectors simply by changing the order of operands. This is because pre-multiplying with a matrix gives the same result as post-multiplying with the transpose of the same matrix.

**Comment by michal-z** [comment]
It works in OpenGL because you are using post-multiplication (you reverse the order of multiplication) so you actually perform a transposition.

In DirectX you have to transpose on the CPU side to use pre-multiplication in the shader. This is because HLSL is column major by default (just like GLSL).

**Comment by presentfactory** [comment]
I guess that makes sense. Suppose they didn't say if they were using DirectX or OpenGL there, just odd though that row major matrices would work natively in OpenGL which is fundamentally column major, but not DirectX without transposition despite zmath having the same convention as DirectXMath...

**Comment by michal-z** [comment]
Additional doc/comment would be very welcome. Would like to make a PR with it?

**Comment by Toqozz** [comment]
Wow, this is confusing.

I've read through all of this (and other resources) many times but wouldn't mind someone sanity checking my logic.

- `zmath` stores matrices in row-major order **in memory**, as evidenced by `translation()` function.
- This is a choice made by `zmath` to align with `DirectXMath`s convention.
- The correct way to transform a vector is `zm.mul(v, m)`, unless you are going against the grain and have made your own matrix which stores the translation in the last column (one translation component in each `zm.f32x4`); then you can use `zm.mul(m, v)` and you'll get a correctly transformed vector.
- All of the samples I've seen in `zig-gamedev` seem to call `transpose()` on the matrices before uploading them to the GPU.  The purpose of this is to maintain the multiplication order and be consistent in the shaders.  **You are not required** to `transpose()` before uploading, but you'd have to switch to post-multiplication instead.
    - For example, you'd have to do `output.pos = object_to_clip * vec4(position, 1.0)` if you didn't `transpose()`, instead of `output.pos = vec4(position, 1.0) * object_to_clip`, which is consistent with `zmath`.

Is this all correct?

This topic is fairly unfriendly because tutorials and such rarely specify why the multiplication order is what it is.  For example, the [Learn WGPU](https://sotrh.github.io/learn-wgpu/beginner/tutorial6-uniforms/#uniform-buffers-and-bind-groups) tutorial for Rust says:
> Multiplication order is important when it comes to matrices. The vector goes on the right, and the matrices go on the left in order of importance.

Which can lead to confusion about whether this is a `zmath` specific thing.

I'm fairly sure that I understand it now, but would be nice to get this concrete, and hopefully this answer can help others as well.  I wouldn't mind contributing to documentation somewhere if you tell me where to put it.
